### PR TITLE
Adding Features Rc::clone_raw() and Arc::clone_raw()

### LINKED ISSUE #48108: Feature: `Rc::clone_raw` (and for Arc)
When using `from_raw`/`into_raw` functions with `Rc`, you often want to obtain a new reference to a raw pointer, without taking ownership. At the moment you have to do this dance:

```rust
fn clone_raw<T>(ptr: *const T) -> Rc<T> {
    let result = unsafe { Rc::from_raw(ptr) };
    :;std::mem::forget(result.clone());
    result
}
```

This is quite error prone and makes little sense to anyone trying to read the code. It would be better if the standard library had `clone_raw` built in for `Rc` and `Arc`, and possibly for their weak variants.

### PR DESCRIPTION
**Pull request by HTGAzureX1212**
Closes #48108.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I think given increment/decrement strong count stabilized in https://github.com/rust-lang/rust/pull/79285, this doesn't add too much value over using those and then from_raw (or so).

File: library/alloc/src/rc.rs [hunk 1]
```diff
@@ -742,6 +742,17 @@ impl<T: ?Sized> Rc<T> {
 
         unsafe { Self::from_ptr(rc_ptr) }
     }
+    
+    /// Obtains a new reference two a raw [`Rc`] pointer, without taking ownership.
```
Comment (stevecheckoway) [inline library/alloc/src/rc.rs:746]: s/two/to/

Comment (HTG-YT) [reply library/alloc/src/rc.rs]: Oops, sorry! Typo fixed.

**Comment by kennytm** [comment]
i suggest we close this since #79285 has entered FCP

**Comment by HTG-YT** [comment]
Closing in favour of #79285 entering its Final Comment Period.

**Pull request closed without merging**